test/e2e/ip_reuse.go

```diff
@@ -0,0 +1,170 @@
+	})
+
+	By("Waiting for all Machines to be Running")
+	WaitForNumMachinesInState(ctx, clusterv1.MachinePhaseRunning, WaitForNumInput{
+		Client:    managementClusterClient,
+		Options:   []client.ListOption{client.InNamespace(input.Namespace)},
+		Replicas:  4,
+		Intervals: input.E2EConfig.GetIntervals(input.SpecName, "wait-machine-remediation"),
+	})
+
+	By("Get the IPPools in the cluster")
+	baremetalv4Pool, provisioningPool := GetIPPools(ctx, managementClusterClient, input.ClusterName, input.Namespace)
+	Expect(baremetalv4Pool).To(HaveLen(1))
+	Expect(provisioningPool).To(HaveLen(1))
+	baremetalv4PoolName := baremetalv4Pool[0].Name
+	provisioningPoolName := provisioningPool[0].Name
+
+	By("Get new values of spec.Preallocations field for baremetal IPPool")
+	bmv4PoolPreallocations, err := GenerateIPPoolPreallocations(ctx, baremetalv4Pool[0], baremetalv4PoolName, managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	Logf("new values of spec.Preallocations field for baremetal IPPool is: %s", bmv4PoolPreallocations)
+	By("Get new values of spec.Preallocations field for provisioning IPPool")
+	provPoolPreallocations, err := GenerateIPPoolPreallocations(ctx, provisioningPool[0], provisioningPoolName, managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	Logf("new values of spec.Preallocations field for provisioning IPPool is: %s", provPoolPreallocations)
+
+	By("Patch baremetal IPPool with new Preallocations field and values")
+	Expect(managementClusterClient.Get(ctx, client.ObjectKey{Namespace: input.Namespace, Name: baremetalv4PoolName}, &baremetalv4Pool[0])).To(Succeed())
+	bmv4helper, err := patch.NewHelper(&baremetalv4Pool[0], managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	baremetalv4Pool[0].Spec.PreAllocations = bmv4PoolPreallocations
+	Expect(bmv4helper.Patch(ctx, &baremetalv4Pool[0])).To(Succeed())
+
+	By("Patch provisioning IPPool with new Preallocations field and values")
+	Expect(managementClusterClient.Get(ctx, client.ObjectKey{Namespace: input.Namespace, Name: provisioningPoolName}, &provisioningPool[0])).To(Succeed())
+	provhelper, err := patch.NewHelper(&provisioningPool[0], managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	provisioningPool[0].Spec.PreAllocations = provPoolPreallocations
+	Expect(provhelper.Patch(ctx, &provisioningPool[0])).To(Succeed())
+
+	// patch MD maxUnavailable and k8s version to start a parallel upgrade
+	By("Get MachineDeployment")
+	machineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+		Lister:      managementClusterClient,
+		ClusterName: input.ClusterName,
+		Namespace:   input.Namespace,
+	})
+	Expect(len(machineDeployments)).To(Equal(1), "Expected exactly 1 MachineDeployment")
+	md := machineDeployments[0]
+
+	Byf("Update MachineDeployment maxUnavailable to number of workers and k8s version from %s to %s", fromK8sVersion, kubernetesVersion)
```

Comment:
Clarification question: I remember introducing intermediary k8s version (since we would have run already k8s upgrade tests before IP Reuse) to just upgrade from one patch version to another, but that was removed. Does that mean we run these tests before those tests?

Reply:
This test will not use any pre existing cluster it will create its own target and use the version needed I used same upgrade from/to we have for k8s upgrade